app/com/arpnetworking/metrics/portal/alerts/impl/CachingAlertExecutionRepository.java

```diff
@@ -0,0 +1,259 @@
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import javax.inject.Named;
+
+/**
+ * An alert repository wrapper that caches the most recent successful results.
+ * <br>
+ * {@link AlertExecutionRepository#getLastSuccessBatch} is probably the most frequently
+ * accessed piece of execution data for alerts, so it's useful to cache those results
+ * separately.
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public final class CachingAlertExecutionRepository implements AlertExecutionRepository {
+    private final AlertExecutionRepository _inner;
+    private final ActorRef _successCache;
+    private final Duration _cacheTimeout;
+
+    private CachingAlertExecutionRepository(final Builder builder) {
+        _inner = builder._inner;
+        _successCache = builder._actorRef;
+        _cacheTimeout = builder._timeout;
+    }
+
+    @Override
+    public void open() {
+        _inner.open();
+    }
+
+    @Override
+    public void close() {
+        _inner.close();
+    }
+
+    @Override
+    public CompletionStage<Optional<JobExecution<AlertEvaluationResult>>> getLastScheduled(
+        final UUID jobId, final Organization organization
+    ) {
+        return _inner.getLastScheduled(jobId, organization);
+    }
+
+    @Override
+    public CompletionStage<Optional<JobExecution.Success<AlertEvaluationResult>>> getLastSuccess(
+        final UUID jobId, final Organization organization
+    ) throws NoSuchElementException {
+        final String key = cacheKey(jobId, organization.getId());
+        return
+            CacheActor.<String, JobExecution.Success<AlertEvaluationResult>>get(_successCache, key, _cacheTimeout)
+                .thenCompose(res -> {
+                    if (res.isPresent()) {
+                        return CompletableFuture.completedFuture(res);
+                    }
+                    return _inner.getLastSuccess(jobId, organization)
+                        .thenCompose(res2 -> {
+                            if (!res2.isPresent()) {
+                                return CompletableFuture.completedFuture(res2);
+                            }
+                            return CacheActor.put(_successCache, key, res2.get(), _cacheTimeout).thenApply(ignore -> res2);
+                        });
+                });
+    }
+
+    @Override
+    public CompletionStage<ImmutableMap<UUID, JobExecution.Success<AlertEvaluationResult>>> getLastSuccessBatch(
+        final List<UUID> jobIds,
+        final Organization organization,
+        final LocalDate maxLookback
+    ) throws NoSuchElementException {
+        // Attempt to get all ids from cache
+        final List<CompletionStage<Optional<JobExecution.Success<AlertEvaluationResult>>>> futures =
```

Comment:
This is expensive. Especially, if it's all remote messages. Should the cache actor support a multi-get?

Reply:
Straightforward so I added it.